frontend/lib/resolution/Resolver.cpp

```diff
@@ -2593,51 +2648,79 @@ void Resolver::resolveTupleUnpackAssign(ResolvedExpression& r,
   }
 
   // Then, check that they have the same size
-  if (lhsTuple->numActuals() != rhsT->numElements()) {
-    context->error(lhsTuple, "tuple size mismatch in split tuple assign");
-    return;
-  }
-  if (lhsT->numElements() != rhsT->numElements()) {
+  if (lhsTuple->numActuals() != rhsT->numElements() ||
+      initialLhsT->numElements() != rhsT->numElements()) {
     context->error(lhsTuple, "tuple size mismatch in split tuple assign");
     return;
   }
 
   const Scope* scope = currentScope();
   auto inScopes = CallScopeInfo::forNormalCall(scope, poiScope);
 
-  // Finally, try to resolve = between the elements
-  int i = 0;
-  for (auto actual : lhsTuple->actuals()) {
+  // Then, make sure that the LHS is valid and adjust its intent.
+  // It recomputes the LHS tuple type and sets it in 'byPostorder'.
+  // It does not recompute intents for component sub-expressions.
+  auto lhsType = adjustLhsForTupleUnpackAssign(*this, astForErr, lhsTuple,
+                                               initialLhsType);
+  auto lhsT = lhsType.type()->toTupleType();
+
+  for (int i = 0; i < lhsTuple->numActuals(); i++) {
+    auto actual = lhsTuple->actual(i);
+
     QualifiedType lhsEltType = lhsT->elementType(i);
     QualifiedType rhsEltType = rhsT->elementType(i);
+    auto ident = actual->toIdentifier();
+
+    // Do not perform an assignment in the case of a '_' variable.
+    if (ident && ident->name() == USTR("_")) continue;
+
     if (auto innerTuple = actual->toTuple()) {
-      resolveTupleUnpackAssign(r, astForErr, innerTuple, lhsEltType, rhsEltType);
-    } else {
-      std::vector<CallInfoActual> actuals;
-      actuals.push_back(CallInfoActual(lhsEltType, UniqueString()));
-      actuals.push_back(CallInfoActual(rhsEltType, UniqueString()));
-      auto ci = CallInfo (/* name */ USTR("="),
-                          /* calledType */ QualifiedType(),
-                          /* isMethodCall */ false,
-                          /* hasQuestionArg */ false,
-                          /* isParenless */ false,
-                          actuals);
-
-      auto c = resolveGeneratedCall(actual, &ci, &inScopes);
-      c.noteResult(&r, { { AssociatedAction::ASSIGN, lhsTuple->id() } });
+      // Recurse if the element is a tuple.
+      resolveTupleUnpackAssign(innerTuple, astForErr, lhsEltType, rhsEltType);
+      continue;
```

Comment:
I find it strange that we handle tuple assignment `=` here, whereas we handle "normal" assignment in `call-init-deinit.cpp`. Is it possible to move this logic there? What about handling tuples constrains this code to run as part of resolution?